src/vm/analysis/type_checker/natives/iterables.rs

```diff
@@ -0,0 +1,210 @@
+        },
+        TypeSignature::BufferType(buffer_data) => {
+            let mapped_type = function_type.check_args(&[TypeSignature::min_buffer()])?;
+            TypeSignature::list_of(mapped_type, buffer_data.into())
+                .map_err(|_| CheckErrors::ConstructedListTooLarge.into())
+        },
+        _ => Err(CheckErrors::ExpectedListOrBuffer(argument_type).into())
+    }
+}
+
+pub fn check_special_filter(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(2, args)?;
+    
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::NonFunctionApplication)?;
+    // we will only lookup native or defined functions here.
+    //   you _cannot_ map a special function.
+    let function_type = get_simple_native_or_user_define(function_name, checker)?;
+        
+    let argument_type = checker.type_check(&args[1], context)?;
+
+    {
+        let input_type = match argument_type {
+            TypeSignature::ListType(ref list_data) => Ok(list_data.clone().destruct().0),
+            TypeSignature::BufferType(_) => Ok(TypeSignature::min_buffer()),
+            _ => Err(CheckErrors::ExpectedListOrBuffer(argument_type.clone()))
+        }?;
+    
+        let filter_type = function_type.check_args(&[input_type])?;
+
+        if TypeSignature::BoolType != filter_type {
+            return Err(CheckErrors::TypeError(TypeSignature::BoolType, filter_type).into())
+        }
+    }
+
+    Ok(argument_type)
+}
+
+pub fn check_special_fold(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(3, args)?;
+    
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::NonFunctionApplication)?;
+    // we will only lookup native or defined functions here.
+    //   you _cannot_ fold a special function.
+    let function_type = get_simple_native_or_user_define(function_name, checker)?;
+    
+    let argument_type = checker.type_check(&args[1], context)?;
+
+    let input_type = match argument_type {
+        TypeSignature::ListType(list_data) => Ok(list_data.destruct().0),
+        TypeSignature::BufferType(_) => Ok(TypeSignature::min_buffer()),
+        _ => Err(CheckErrors::ExpectedListOrBuffer(argument_type))
+    }?;
+
+    let initial_value_type = checker.type_check(&args[2], context)?;
+
+    // fold: f(A, B) -> A
+    //     where A = initial_value_type
+    //           B = list items type
+    
+    // f must accept the initial value and the list items type
+    let return_type = function_type.check_args(&[input_type.clone(), initial_value_type])?;
+
+    // f must _also_ accepts its own return type!
+    let return_type = function_type.check_args(&[input_type, return_type])?;
+    
+    Ok(return_type)
+}
+
+pub fn check_special_concat(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
```

Comment:
Not sure if this is the intended design, but we're saying that it is _impossible_ to concatenate a `list` and `buff` or a `buff` to a `list`, right?  If that's the case, then is there a way to convert a `list` to a `buff` and back so I can concatenate them safely?

Reply:
> Not sure if this is the intended design, but we're saying that it is impossible to concatenate a list and buff or a buff to a list

Correct. 

> is there a way to convert a list to a buff and back so I can concatenate them safely?

A list can contain a collection of int, uint, but also more complex constructs like tuples or tuples of tuples.
I think the first step for supporting the feature you're asking for would be being able to serialize `any` into a `buff`. Then from there you would be able to `map` / `reduce` a list and turn it into one concatenated buffer.
Does that make sense? I'll let @kantai chime in if it does not :)

Reply:
We might want to consider adding a native function for converting a `buff` into a `list`, but I'm fine with the code as-is if it's not too hard to write a simple Clarity routine for converting a `buff` to a `list` and back (assuming the `list` contains appropriately-sized `uint`s).

Reply:
Part of the issue is that there's no "atomic" element of buffers in the language currently. When you `map` `fold` or `filter` a buffer, you are operating over types `(buff 1)`, not `uint`s.

Reply:
Ah, I did not know that.  I would have expected that `map`, `fold`, or `filter` would have acted on the `buff`'s elements as if they were `uint`.  I think this raises a deeper question -- to what extent do we want programmers to think of `buff` as list-like?  We don't have to answer this in this PR, but I think we'll want to know for sure when we close #1056.

Reply:
Yep -- I suspect if we did want buffers to be more like `list`s we'd want a different type for the buff element (`char` or `u8`) than we have for integers currently (`u128`/`i128`).